File: app/src/main/java/com/sies/movierecomendations/SearchActivity.java
package com.sies.movierecomendations;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.speech.RecognizerIntent;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.widget.EditText;
import android.widget.ImageView;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.sies.movierecomendations.MoviesApi.MoviesList;

import java.util.ArrayList;
import java.util.Locale;
import java.util.Objects;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class SearchActivity extends AppCompatActivity {

    EditText searchBar;
    RecyclerView recyclerView;
    Toolbar toolbar;
    ImageView voiceSearch;
    String key = BuildConfig.API_KEY;
    MoviesList res;

    Retrofit retrofit = new Retrofit.Builder()
            .baseUrl("https://api.themoviedb.org/3/")
            .addConverterFactory(GsonConverterFactory.create())
            .build();
    MovieDbAPI movieDbAPI = retrofit.create(MovieDbAPI.class);

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_search);

        searchBar = findViewById(R.id.searchBar);
        recyclerView = findViewById(R.id.recyclerView);
        toolbar = findViewById(R.id.toolbar);
        voiceSearch = findViewById(R.id.voiceSearch);

        searchBar.requestFocus();

        voiceSearch.setOnClickListener(v -> voiceSearch());

        setSupportActionBar(toolbar);
        Objects.requireNonNull(getSupportActionBar()).setDisplayHomeAsUpEnabled(true);

        recyclerView.setLayoutManager(new LinearLayoutManager(SearchActivity.this));

        searchBar.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if(count>0)
                    getApi(s.toString());
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
    }
    public void getApi(String s) {
        movieDbAPI.getSearchResult(key, s).enqueue(new Callback<MoviesList>() {
            @Override
            public void onResponse(@NonNull Call<MoviesList> call, @NonNull Response<MoviesList> response) {
                res = response.body();
                recyclerView.setAdapter(new SearchAdapter(res));
            }

            @Override
            public void onFailure(@NonNull Call<MoviesList> call, @NonNull Throwable t) {
                Log.i("onFailure: ", t.getMessage());
            }
        });
    }

    ActivityResultLauncher<Intent> voiceActivityResultLauncher = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(), result -> {
                if (result.getResultCode() == Activity.RESULT_OK) {
                    // There are no request codes
                    Intent data = result.getData();
                    ArrayList<String> text;
                    if (data != null) {
                        text = data.getStringArrayListExtra(RecognizerIntent.EXTRA_RESULTS);
                        searchBar.setText(text.get(0));
                    }
                }
            });

    public void voiceSearch() {
        Intent intent = new Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH);
        intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM);
        intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE, Locale.getDefault());

        if (intent.resolveActivity(getPackageManager()) != null)
            voiceActivityResultLauncher.launch(intent);
    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }
}

File: app/src/main/java/com/sies/movierecomendations/Search/ui/SearchActivity.kt
package com.sies.movierecomendations.Search.ui

import android.content.Intent
import android.os.Bundle
import android.speech.RecognizerIntent
import android.text.Editable
import android.text.TextWatcher
import androidx.activity.result.contract.ActivityResultContracts.StartActivityForResult
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProvider
import com.sies.movierecomendations.MovieDetails.ui.MovieDetails
import com.sies.movierecomendations.R
import com.sies.movierecomendations.Search.data.SearchViewModel
import com.sies.movierecomendations.databinding.ActivitySearchBinding
import java.util.*

class SearchActivity : AppCompatActivity() {

    lateinit var binding: ActivitySearchBinding
    lateinit var viewModel: SearchViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_search)
        binding.lifecycleOwner = this

        viewModel = ViewModelProvider(this).get(SearchViewModel::class.java)

        val adapter = SearchAdapter(SearchItemClickListener {
            viewModel.passDetailsToAdapter(it)
        })

        binding.recyclerView.adapter = adapter
        viewModel.result.observe(this, { it?.let {
            adapter.submitList(it.results)
        }})

        viewModel.term.observe(this, {
            it?.let {
                val intent = Intent(this, MovieDetails::class.java)
                if (it.media_type == "tv"){
                    intent.putExtra("title", it.original_name)
                    intent.putExtra("release-date", it.first_air_date)
                }
                else {
                    intent.putExtra("title", it.title)
                    intent.putExtra("release-date", it.release_date)
                }
                intent.putExtra("backdrop-path", it.backdrop_path)
                intent.putExtra("rating", it.vote_average)
                intent.putExtra("overview", it.overview)
                intent.putExtra("id", it.id)
                startActivity(intent)
            }
        })

        binding.searchBar.requestFocus()
        binding.voiceSearch.setOnClickListener { voiceSearch() }

        setSupportActionBar(binding.toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        binding.searchBar.addTextChangedListener(object: TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                if (count > 0) viewModel.search(s.toString())
            }
            override fun afterTextChanged(s: Editable) {}
        })
    }

    private var voiceActivityResultLauncher = registerForActivityResult(StartActivityForResult()) {
        if (it.resultCode == RESULT_OK) {
            // There are no request codes
            val data = it.data
            val text: ArrayList<String>?
            if (data != null) {
                text = data.getStringArrayListExtra(RecognizerIntent.EXTRA_RESULTS)
                binding.searchBar.setText(text?.get(0))
            }
        }
    }

    private fun voiceSearch() {
        val intent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH)
        intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM)
        intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE, Locale.getDefault())
        if (intent.resolveActivity(packageManager) != null)
            voiceActivityResultLauncher.launch(intent)
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return true
    }
}